internal/keys/hasher.go

```diff
@@ -68,9 +79,24 @@ func (t tupleKeysHasher) Append(h hasher) error {
 
 	n := 0
 	for _, tupleKey := range sortedTupleKeys {
-		key := fmt.Sprintf("%s#%s@%s", tupleKey.GetObject(), tupleKey.GetRelation(), tupleKey.GetUser())
+		key := strings.Builder{}
+		key.WriteString(tupleKey.GetObject())
+		key.WriteString("#")
+		key.WriteString(tupleKey.GetRelation())
+		key.WriteString("@")
+		key.WriteString(tupleKey.GetUser())
+
+		cond := tupleKey.GetCondition()
+		if cond != nil {
+			key.WriteString(cond.GetName())
```

Comment:
**Important:** We need to have a separator between condition and the preceding string. Otherwise, cache collisions will occur. For example, if you have `user:is_` with condition `grant_ok` the cache key would be the same as `user:is_grant_ok` without a condition. 

We'll want to separate with a character that is forbidden from the user field (which we may or not have based on [api spec](https://github.com/openfga/api/blob/24dec682b31d2864ca016f85f5c1a1918e92fd87/openfga/v1/openfga.proto#L161)).

Reply:
From the spec it looks like we don't restrict `user` at all. Object and Relation each have patterns with disallowed characters, but not user. What do you think we should do here? `#` seems like a safe bet, but that already has meaning in FGA. Some other special character?

Reply:
I'm not sure we need one there, as these cache keys are never used in any way that requires them to be readable. They just have to be unique so that their [eventual uint conversion](https://github.com/openfga/openfga/blob/main/pkg/storage/cache.go#L196) results in a unique integer.

The existing separators are here to prevent overlap cases, e.g. `document:reviewer` vs `document:re#viewer`, but the combination of condition name and context hash will always be unique (since the context hasher sorts keys before writing).

Reply:
Update: I took a look back over this today and there _is_ an unaccounted for edge case which your suggestion would address 👍 . I'll add another separator

Reply:
Update 2: I've set this to auto-merge as-is once CI passes without adding the additional separator. I plan on adding it in a quick follow up PR, but doing so requires a refactor of how our hasher behaves and I don't want the blast radius on this PR to be any bigger than it already is. I will link that PR to this one when it's up